### ISSUE TITLE
Diff script ignores number of deleted or changed files

### ISSUE THREAD

**Issue by metrip** [OpenMediaVault-Plugin-Developers/openmediavault-snapraid#24]
`omv-snapraid-diff` ignores number of deleted or changed files.

Introduced in 05d009fe8ccef1468562936e9eff5ea8550699c8 by removing `else`s.

Sample log with 67 deleted files, threshold configured to 50

```
[2020-03-07 00:00:35] omv-snapraid-diff: INFO: 'SnapRAID DIFF finished'
[2020-03-07 00:00:35] omv-snapraid-diff: INFO: '----------------------------------------'
[2020-03-07 00:00:35] omv-snapraid-diff: INFO: 'SUMMARY of changes since last sync:'
[2020-03-07 00:00:35] omv-snapraid-diff: INFO: 'Added: [1] - Deleted: [67] - Moved: [0] - Copied: [0] - Updated: [0]'
[2020-03-07 00:00:35] omv-snapraid-diff: INFO: 'Changes detected [A-1,D-67,M-0,C-0,U-0] -> there are updated files (0) but update threshold (0) is disabled. Running SYNC Command.'
[2020-03-07 00:00:35] omv-snapraid-diff: INFO: 'Number of deleted files (67) exceeded threshold (50). NOT proceeding with sync job.'
[2020-03-07 00:00:35] omv-snapraid-diff: INFO: 'Please run sync manually if this is not an error condition.'
[2020-03-07 00:00:35] omv-snapraid-diff: INFO: 'Full statusreport sent to '''
[2020-03-07 00:00:35] omv-snapraid-diff: INFO: 'Changes detected [A-1,D-67,M-0,C-0,U-0] -> deleted files (67) is below threshold (50). Running SYNC Command.'
```

**Comment by ryecoaaron** [comment]
Can you explain why removing else statements that performed nothing (just had a comment in the block) caused this?

**Comment by metrip** [comment]
The else statements were not empty blocks - they were closed by the corresponding `fi`. They controlled when the code reached the part doing the sync command. The varying indentation makes this unclear.

Now, the sync command is reached every time, regardless of the number of updated and deleted files.

The 1.0.9 version also seems to be missing one `fi` statement.

**Comment by ryecoaaron** [comment]
My bad.  Just another reason why I shouldn't touch this script.  It has so many things that need to be fixed and I don't use snapraid.  So, I'm not the right one to fix.  Feel free to submit a pull request fixing it.

**Comment by auanasgheps** [comment]
@ryecoaaron I am using a great script which could be a perfect candidate. It works perfectly as is and has a lot of customization options (some of them are already built in the GUIO

**Comment by ryecoaaron** [comment]
I write the plugin but solo815 on the forum helped add the script.  He hasn't been around in a long time though.

**Comment by auanasgheps** [comment]
Allright.
 Since I don't know how to make my script working with the plugin GUI, feel free to contact me when somebody shows up with such skills.

**Comment by ryecoaaron** [comment]
You just asked who did the work on the current script.  Since I wrote the plugin and know scripting very well, I have the skills.  I just don't use snapraid and don't have the time to spend to learn everything.

**Comment by auanasgheps** [comment]
I got lost in translation... 😅 Perfect, I will create a separate issue for this.

**Comment by jpinnekamp** [comment]
@ryecoaaron:
Since this issue is still open, I'd like to point out that there is a pull request waiting to be merged (actually since before this issue was opened).

**Comment by ryecoaaron** [comment]
@jpinnekamp I don't have a good way to test your changes.  If a couple of people could test them, I would merge it.

**Comment by jpinnekamp** [comment]
The changes are not actually mine, just happened to find the PR. Sorry not making that clear.
I'm not an expert at bash myself, but upon review it looks good. Will test it out and report back.

**Comment by jpinnekamp** [comment]
@ryecoaaron: So it works for me, even though the emails are a bit to verbose for my taste. Not sure if that part is a change to previous behavior though.

**Comment by sualfred** [comment]
Noticed this recently with omv4 and now with my new server on omv5.

```Changes detected [A-8,D-53119,M-0,C-0,U-92] -> deleted files (53119) is below threshold (500). Running SYNC Command```

@jpinnekamp have switched to the PR script (this PR has been open since 2019 and it's not a minor bug?!). Thanks in advance. I'll report back if I notice issues or if it's working

**Comment by auanasgheps** [comment]
If you are looking for a good script, I [made one](https://github.com/auanasgheps/snapraid-aio-script). It's pretty mature and almost feature complete. 
It would be a great addition to the plugin, but I'm sure a lot of work would be required to integrate it, since there's also the UI/Plugin to be adapted.

**Comment by ryecoaaron** [comment]
@sualfred Yes it has been open that long but I said I needed a couple of people to test it before I would merge.  I did not have feedback from more than one person.  Bug or not, I still need multiple testers on a plugin used by so many.

@auanasgheps I don't have any problem switching the script and make the plugin changes IF someone maintains the script.  I don't want to do that for OMV 5 but I would for OMV 6 since I started porting the plugin.

**Comment by sualfred** [comment]
@ryecoaaron 
As I said, I'll provide feedback asap. Since this is a critical bug which could cause data loss the PR is a way open to long. I really don't understand why a popular NAS system has so much users who don't read their report emails of snapraid ^^

**Comment by ryecoaaron** [comment]
@sualfred How would it cause data loss?  Snapraid shouldn't delete files from the filesystem.  Worst case you restore from backup since snapraid isn't backup anymore than raid is.  But if you find the merge request fixes the issue, I will merge it.

**Comment by auanasgheps** [comment]
> @auanasgheps I don't have any problem switching the script and make the plugin changes IF someone maintains the script. I don't want to do that for OMV 5 but I would for OMV 6 since I started porting the plugin.

I'm here to maintain my own script so don't worry about this, although my bash capabilities are limited :)

I've done most of the stuff I planned, but it would be great if you could just "plug" the UI/Plugin into the script so subsequent updates would be hassle free. (I'm planning a small, cosmetic update soon).

I'm guessing it could be done since the configuration is stored in a separate file.

**Comment by sualfred** [comment]
@ryecoaaron 
I'm far away from a snapraid expert. But isn't it true that I cannot recover deleted files after the sync?
That would mean that a stupid unwanted and not noticed deletion or any other data loss cannot be restored.

**Comment by auanasgheps** [comment]
> @ryecoaaron
> I'm far away from a snapraid expert. But isn't it true that I cannot recover deleted files after the sync?
> That would mean that a stupid unwanted and not noticed deletion or any other data loss cannot be restored.

True, but you should not use SnapRAID as a backup system. The threshold is there to protect against  ransomware or catastrophic involuntary deletions. You should treat SnapRAID as a raid solution. Do your backups properly

**Comment by sualfred** [comment]
Sure, true. For critical data. 

But a RAID is a redundant array to one logical drive. Snapraid != Raid. Snapraid is a snaptshot system and has nothing to do with building an array. It's a very simple and good backup solution if one or two drives fail or data has been lost. That's why this is a critical bug. 

Don't get me wrong. This general discussion and saying that "do you backups properly" is super annoying because there is no black and white.

Long story short, this is a critical bug because it CAN cause data loss and the PR is working. That's all I can say :)

**Comment by ryecoaaron** [comment]
> But isn't it true that I cannot recover deleted files after the sync?

Yes but this is only data loss if you have a drive failure.  If you have two drives fail and you only have one parity drive, you have data loss as well.

And snapraid is raid.  It just isn't realtime raid.  It calculates parity just like raid.
